pkg/volume/volume_linux.go

```diff
@@ -40,27 +41,96 @@ const (
 // SetVolumeOwnership modifies the given volume to be owned by
 // fsGroup, and sets SetGid so that newly created files are owned by
 // fsGroup. If fsGroup is nil nothing is done.
+var (
+	recorder record.EventRecorder
+	pod      *v1.Pod
+)
+
 func SetVolumeOwnership(mounter Mounter, dir string, fsGroup *int64, fsGroupChangePolicy *v1.PodFSGroupChangePolicy, completeFunc func(types.CompleteFuncParam)) error {
+
 	if fsGroup == nil {
 		return nil
 	}
 
 	timer := time.AfterFunc(30*time.Second, func() {
-		klog.Warningf("Setting volume ownership for %s and fsGroup set. If the volume has a lot of files then setting volume ownership could be slow, see https://github.com/kubernetes/kubernetes/issues/69699", dir)
+		klog.Warningf("Setting volume ownership for %s and fsGroup set. If the volume has a lot of files, setting volume ownership could be slow, see https://github.com/kubernetes/kubernetes/issues/69699", dir)
 	})
+
 	defer timer.Stop()
 
 	if skipPermissionChange(mounter, dir, fsGroup, fsGroupChangePolicy) {
 		klog.V(3).InfoS("Skipping permission and ownership change for volume", "path", dir)
 		return nil
 	}
 
+	// Channel for progress updates, using buffered channel for deadlock prevention.
+	progressChannel := make(chan ProgressUpdate, 1)
+
+	// Goroutine for handling progress updates
+	go func() {
+		for update := range progressChannel {
+			klog.V(2).Infof("FSGroup permission change in progress for %s: %d files, %d directories processed (elapsed time: %v)", dir, update.Files, update.Dirs, update.Elapsed)
+
+			// Emit event.
+			if pod != nil && recorder != nil {
+				recorder.Eventf(pod, v1.EventTypeNormal, "FSGroupChangeProgress",
+					"FSGroup permission change in progress: %d files, %d directories processed (elapsed time: %v)",
+					update.Files, update.Dirs, update.Elapsed)
+			} else {
+				klog.Warningf("Pod or recorder is nil, skipping event emission")
+			}
+		}
+	}()
+
+	// File counters and time tracking
+	var fileCount, dirCount int
+	startTime := time.Now()
 	err := walkDeep(dir, func(path string, info os.FileInfo, err error) error {
 		if err != nil {
+			klog.Errorf("Error accessing path %s: %v", path, err)
+			return err
+		}
+
+		// Apply permission changes
```

Comment:
For manual testing, maybe we can change this time interval to 1 seconds and try to see the event and log is recorded correctly.